01-resources-v2/proposal.md

```diff
@@ -0,0 +1,412 @@
+# Summary
+
+Introduces a new resource interface with the following goals:
+
+* Introduce versioning to the resource interface, so that we can maintain
+  backwards-compatibility.
+
+* Support for spaces (concourse/concourse#1707).
+
+* Introduce a more airtight "versioned artifacts" interface, tightening up
+  loopholes in today's resource API to ensure that resources are pointing at an
+  external source of truth, so that we can explicitly design for new features
+  and workflows rather than forcing everything into the resource interface.
+
+* Extend the versioned artifacts interface to support deletion of versions.
+
+
+# Proposal
+
+## General Types
+
+```go
+// Space is a name of a space, e.g. "master", "release/3.14", "1.0".
+type Space string
+
+// Config is a black box containing all user-supplied configuration, combining
+// `source` in the resource definition with `params` from the step (in the
+// case of `get` or `put`).
+type Config map[string]interface{}
+
+// Version is a key-value identifier for a version of a resource, e.g.
+// `{"ref":"abcdef"}`, `{"version":"1.2.3"}`.
+type Version map[string]string
+
+// Metadata is an ordered list of metadata fields to display to the user about
+// a resource version. It's ordered so that the resource can decide the best
+// way to show it.
+type Metadata []MetadataField
+
+// MetadataField is an arbitrary key-value to display to the user about a
+// version of a resource.
+type MetadataField struct {
+  Name  string `json:"name"`
+  Value string `json:"value"`
+}
+```
+
+## Versioned Artifacts interface
+
+### `check`: Detect versions across spaces.
+
+The `check` command will be invoked with the following JSON structure on
+`stdin`:
+
+```go
+// CheckRequest contains the resource's configuration and latest version
+// associated to each space.
+type CheckRequest struct {
+  Config Config            `json:"config"`
+  From   map[Space]Version `json:"from"`
+}
+```
+
+The first call will have an empty object as `from`.
+
+Any spaces discovered by the resource but not present in `from` should collect
+versions from the beginning.
```

Comment:
I'm still going through the performance implications of collecting *every* version for *every* space for *every* resource in *every* pipeline. This is what I'm most cautious about.

## Disk usage on DB

Collecting all the metadata for every version impacts disk usage on the database. I mean...that's just a fact, but I don't think it's a dealbreaker. This usage should be dwarfed by the space consumed by build logs, both in volume and rate of increase.

Mitigation: we may be able to share data between teams/pipelines by associating the versions/metadata to the global "resource cache" objects rather than per-pipeline entities. Even if we don't, I still think this shouldn't be a huge deal.


## Time/CPU/RAM consumption on workers

How do we emit the versions and metadata for, say, 700k commits to a monolithic git repo? How often will resource authors mistakenly implement this by holding everything in RAM?

Mitigation: I've added `has_latest: true` to the response, which will be required for Concourse to consider the versions to be "fully collected" and start using the versions for pipeline scheduling. This makes batching the versions/metadata held in RAM fairly easy (checking always goes chronologically and can be done with arbitrary batch sizes at the resource type author's discretion). Downside: it will take time to collect everything on first pipeline configuration. We can check spaces that don't have their latest version more aggressively until everything is collected.

## Scheduling performance

The current approach to pipeline scheduling works by loading literally every version of every resource into memory. This at least only loads database IDs (no version strings and certainly no metadata), but the sheer volume of versions to walk through and try when finding input candidates could really clog things up.

Mitigation: we could modify the scheduling algorithm to only load the latest 1000 or so versions of each resource at a time, and lazy-load older ones as needed (if the given set yields no input candidates, load 1000 more of everything). This should speed up scheduling efficiency in most cases anyway (it should be very infrequent that the latest valid candidate is a very old version), and reduce the memory usage on the `web` node overall (because the versions db is cached per-pipeline).

@itsdalmo Any thoughts/concerns on this?

Reply:
Hey!

> How often will resource authors mistakenly implement this by holding everything in RAM?

I think I'd be guilty of holding everything in memory myself, but from your tests it seems like this would be fine for everything but the edge cases (which I consider the Linux repo to be). And even with the linux repo the absolute memory usage is not that crazy, but perhaps the memory usage relative to other resources might trip people up?

> Mitigation: I've added has_latest: true to the response, which will be required for Concourse to consider the versions to be "fully collected" and start using the versions for pipeline scheduling.

So, I can e.g. paginate an API and emit versions, and only when I get to the last page and emit `has_latest: true` will Concourse schedule a task which has `trigger: true`? Is this only for the initial check, or for all checks? If so, how does it work with `every: true` (and/or ignore it on the initial check)?

Overall I think this is a good mitigation which makes it a lot easier to write lightweight resources. One way of _ensuring_ that resources are written in a lightweight manner might be to say that each version should be written individually to stdout, so resource authors would be encouraged to emit versions straight from a for-loop, rather than growing a temporary array.

What would be the drawbacks of doing that? We'd need to include the `space` for every version (rather than every array), and all the flushes to stdout might degrade performance?

> Mitigation: we could modify the scheduling algorithm to only load the latest 1000 or so versions of each resource at a time, and lazy-load older ones as needed (if the given set yields no input candidates, load 1000 more of everything). This should speed up scheduling efficiency in most cases anyway (it should be very infrequent that the latest valid candidate is a very old version), and reduce the memory usage on the web node overall (because the versions db is cached per-pipeline).

Perhaps I'm misunderstanding this, but in relation to `every: true` - it sounds like  it would either have to continuously loop over all versions to ensure that none have been missed, or that it would not be guaranteed to build every version? If the latter is the case, then I think the behaviour needs to be documented, or ideally that `every: true` should instead be `every: N` (where N could be 1000) or something to that effect?

Reply:
> And even with the linux repo the absolute memory usage is not that crazy, but perhaps the memory usage relative to other resources might trip people up?

It can actually get pretty high because we're returning all the metadata now (which includes commit messages). I believe it got up to a few gigabytes in my testing. The windowing, plus the first parent walking simplification, helps dramatically.

> So, I can e.g. paginate an API and emit versions, and only when I get to the last page and emit has_latest: true will Concourse schedule a task which has trigger: true? Is this only for the initial check, or for all checks? If so, how does it work with every: true (and/or ignore it on the initial check)?

I think it applies to all checks. It'll apply to all jobs, and it shouldn't matter whether `trigger: true` is set because that just influences whether Concourse auto-queues a build, not how the candidates are determined.

Here's how I think it'd work:

1. Run `check`, get `has_latest: false`. Save the versions, with a `partial` column set to `true` in the database. (This is a new column I just made up that could maybe use a better name.)
1. Modify the scheduler to only load versions with `partial` as false.
1. Run `check` again, this time it gets `has_latest: true`. Save the versions, and then set `partial` to `false` for all versions of the space.
1. Next time the scheduler runs, the full batch of versions will now be available.

This logic applies to both the initial check (because it may take a while to collect everything) and any subsequent checks (because a ton of versions may appear at once, and people would be annoyed if we started running from some random commit in the middle).

Also this should have no impact on `version: every` (assuming that's what you mean). It's just to prevent windows of time where we have partially saved data that would otherwise result in running things the user doesn't want.

> Overall I think this is a good mitigation which makes it a lot easier to write lightweight resources. One way of ensuring that resources are written in a lightweight manner might be to say that each version should be written individually to stdout, so resource authors would be encouraged to emit versions straight from a for-loop, rather than growing a temporary array.
> 
> What would be the drawbacks of doing that? We'd need to include the space for every version (rather than every array), and all the flushes to stdout might degrade performance?

I considered this, and it's really tempting, but I don't quite trust Garden's process output streaming enough. If the process emits enough output so quickly that the consumer can't keep up, Garden will begin dropping chunks of output so that the server doesn't get backpressured. Come to think of it, that may be reason enough to not use `stdout` for the response at all, now that it's going to be much larger. We might have to change the interface so that `check` writes the response to a file. (This would at least match the behavior of `put`.)

@topherbullock Would writing the resource's response to a file in the container make k8s support (concourse/rfcs#2) any easier, btw? (As opposed to streaming responses on `stdout`.)

> Perhaps I'm misunderstanding this, but in relation to every: true - it sounds like it would either have to continuously loop over all versions to ensure that none have been missed, or that it would not be guaranteed to build every version? If the latter is the case, then I think the behaviour needs to be documented, and perhaps something users can tweak on the resource configuration?

Good catch, I haven't thought about how that would work. `version: every` is currently implemented by [checking if the prior version has been used](https://github.com/concourse/atc/blob/9b578913f7f273197efd0fb1866558da88d4e74d/db/algorithm/input_candidates.go#L24-L47) - we would have to ensure this doesn't get tripped up by the windowing on the `.Peek` call.

Reply:
Follow-up: writing the response to a file or stdout doesn't make it any easier or harder for k8s. I went with writing to a file for consistency with `put` and safety (since we can't 100% trust the streaming process output).